File: c/atbash-cipher/atbash_cipher.c
#include "atbash_cipher.h"
#include <assert.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>

typedef enum {
  DECODE,
  ENCODE
} CODE_TYPE;

static void atbash_code(const char *input, char **output, CODE_TYPE side){
  int length = strlen(input);
  int i = 0; // code index
  int space_divider = 5;
  for (int index = 0; index < length; index++){
    char ch = input[index];
    if (!isalnum(ch))
      continue;

    // space ' ' workaround
    if (side == ENCODE && i && i % space_divider == 0) {
      (*output)[i++] = ' ';
      space_divider += 6;
    }  
    (*output)[i++] = isalpha(ch) ? 'z' + 'a' - tolower(ch) : ch;
  }
}

char *atbash_encode(const char *input){
  assert(input);
  int length = strlen(input);
  char * code = calloc(length * 1.2 + 1, sizeof(char)); // + 20% causes of spaces
  if (!code) return NULL;

  atbash_code(input, &code, ENCODE);
  code = realloc(code, (strlen(code) + 1) * sizeof(char));

  return code;
}

char *atbash_decode(const char *input){
  assert(input);
  int length = strlen(input);
  char * text = calloc(length + 1, sizeof(char));
  if (!text) return NULL;

  atbash_code(input, &text, DECODE);
  text = realloc(text, (strlen(text) + 1) * sizeof(char));

  return text;
}

File: python/atbash-cipher/atbash_cipher.py
from string import ascii_lowercase


def encode(plain_text):
    return ''.join([('' if index % 5 else ' ') + ll
                    for index, ll in enumerate(code(plain_text))]).strip()


def decode(ciphered_text):
    return code(ciphered_text)


def code(text):
    c_map = build_cipher_map()
    return ''.join([(c_map[ll] if ll.isalpha() else ll)
                    for ll in text.lower() if ll.isalnum()])


def build_cipher_map():
    return dict(zip(ascii_lowercase, ascii_lowercase[::-1]))
